src/biogeochem/Fan2CTSMMod.F90

```diff
@@ -0,0 +1,1059 @@
+    real(r8), intent(in) :: tan_fract_excr
+    integer                  , intent(in)    :: num_soilc       ! number of soil columns in filter
+    integer                  , intent(in)    :: filter_soilc(:) ! filter for soil columns
+
+    integer :: begg, endg, g, l, c, il, counter, col_grass, status, p
+    real(r8) :: flux_avail_rum, flux_avail_mg, flux_grazing, invscale
+    real(r8) :: tempr_ave, windspeed_ave ! windspeed and temperature averaged over agricultural patches
+    real(r8) :: tempr_barns, tempr_stores, vent_barns, flux_grass_crop, tempr_min_10day, &
+         flux_grass_graze, flux_grass_spread, flux_grass_spread_tan, flux_grass_crop_tan
+    real(r8) :: cumflux, totalinput, total_to_store, total_to_store_tan
+    ! dimensions are (type of flux, ruminant/other)
+    real(r8) :: fluxes_nitr(num_fluxes,2), fluxes_tan(num_fluxes,2) 
+    ! The fraction of manure applied continuously on grasslands (if present in the gridcell)
+    real(r8), parameter :: kg_to_g = 1e3_r8
+    logical :: is_grass
+    
+    begg = bounds%begg; endg = bounds%endg
+    nh3_flux_stores(bounds%begc:bounds%endc) = 0_r8
+    nh3_flux_barns(bounds%begc:bounds%endc) = 0_r8
+    man_n_barns(bounds%begc:bounds%endc) = 0.0_r8
+    
+    totalinput = 0.0
+    cumflux = 0.0
+    
+    do g = begg, endg
+       ! First find out if there are grasslands in this cell. If yes, a fraction of
+       ! manure can be diverted to them before storage.
+       col_grass = ispval
+       do il = 1, max_lunit
+          l = grc%landunit_indices(il, g)
+          if (l == ispval) cycle
+          if (lun%itype(l) == istsoil) then
+             do p = lun%patchi(l), lun%patchf(l)
+                is_grass = patch%itype(p) == nc4_grass &
+                     .or. patch%itype(p) == nc3_nonarctic_grass &
+                     .or. patch%itype(p) == nc3_arctic_grass
+                if (is_grass .and. col%wtgcell(patch%column(p)) > 1e-6) then
+                   col_grass = patch%column(p)
+                   exit
+                end if
+             end do
+          end if
+          if (col_grass /= ispval) exit
+       end do
+       ! Transfer of manure from all crop columns to the natural vegetation column:
+       flux_grass_graze = 0_r8
+       flux_grass_spread = 0_r8
+       flux_grass_spread_tan = 0_r8
+
+       do il = 1, max_lunit
+          l = grc%landunit_indices(il, g)
+          if (l == ispval) cycle
+          if (lun%itype(l) == istcrop) then
+             ! flux_avail = manure excreted per m2 of crops (ndep_mixed_grc = per m2 / all land units)
+             do c = lun%coli(l), lun%colf(l)
+                if (.not. col%active(c)) cycle
+                if (col%wtgcell(c) < 1e-6) cycle
+
+                if (col%landunit(c) /= l) then
+                   write(iulog, *) g, il, c, col%landunit(c)
+                   call endrun('something wrong')
+                end if
+                if (.not. any(c==filter_soilc(1:num_soilc))) then
```

Comment:
This (checking, for every column, if the column is contained in the entire filter array) is a very expensive operation that should be reworked.

Reply:
This check was certainly overkill for production code.